tree/treeplayer/inc/ROOT/TVec.hxx

```diff
@@ -0,0 +1,381 @@
+
+} // End of Experimental NS
+
+} // End of VecOps NS
+
+// Other helpers
+namespace Internal {
+
+namespace VecOps {
+
+using namespace ROOT::Experimental::VecOps;
+
+void CheckSizes(size_t s0, size_t s1, const char *opName)
+{
+   if (s0 != s1) {
+      std::stringstream err;
+      err << "Cannot perform operation " << opName
+          << ". The array sizes differ: "
+          << s0 << " and " << s1 << std::endl;
+      throw std::runtime_error(err.str());
+   }
+}
+
+template <typename T, typename V, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, const TVec<V> &v1, const char* opName, F f)
+{
+   CheckSizes(v0.size(), v1.size(), opName);
+   TVec<typename TCallTraits<F>::ret_type> w;
+   w.resize(v0.size());
+   std::transform(v0.begin(), v0.end(), v1.begin(), w.begin(), f);
+   return w;
+}
+
+template <typename T, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, F f)
+{
+   TVec<typename TCallTraits<F>::ret_type> w;
+   w.resize(v0.size());
+   std::transform(v0.begin(), v0.end(), w.begin(), f);
+   return w;
+}
+
+
+template <typename...>
+struct TIsOneOf {
+   static constexpr bool value = false;
+};
+
+template <typename F, typename S, typename... T>
+struct TIsOneOf<F, S, T...> {
+   static constexpr bool value = std::is_same<F, S>::value || TIsOneOf<F, T...>::value;
+};
+
+template <typename T>
+struct TIsChar {
+   static constexpr bool value =
+      TIsOneOf<typename std::decay<T>::type, char, signed char, unsigned char, wchar_t, char16_t, char32_t>::value;
+};
+
+} // End of VecOps NS
+
+} // End of Internal NS
+
+
+namespace Detail {
+
+namespace VecOps {
+
+   template <typename>
+   struct TIsTVec {
+      static const bool value = false;
+   };
+
+   template <typename T>
+   struct TIsTVec<ROOT::Experimental::VecOps::TVec<T>> {
+      static const bool value = true;
+   };
+
+} // End of VecOps NS
+
+} // End of Detail NS
+
+namespace Experimental {
+
+namespace VecOps {
+
+/** @name Math Operators with scalars
+ *  Math operators involving TVec
+*/
+///@{
+template <typename T, typename V>
+auto operator+(const TVec<T> &v, const V &c) -> TVec<decltype(v[0] + c)>
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) {return t+c;});
+}
+
+template <typename T, typename V>
+auto operator-(const TVec<T> &v, const V &c) -> TVec<decltype(v[0] - c)>
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) {return t-c;});
+}
+
+template <typename T, typename V>
+auto operator*(const TVec<T> &v, const V &c) -> TVec<decltype(v[0] * c)>
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) {return t*c;});
+}
+
+template <typename T, typename V>
+auto operator/(const TVec<T> &v, const V &c) -> TVec<decltype(v[0] / c)>
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) {return t/c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
```

Comment:
I acknowledge all comments about the usage of char as boolean type, both coming from the programming model point of view and the optimisation one. This will be changed.